src/Stan/Pattern/Ast.hs

```diff
@@ -81,11 +84,69 @@ infixr 7 ***, +++, ???
 fixity :: PatternAst
 fixity = PatternAstNode $ one ("FixitySig", "FixitySig")
 
-{- | Pattern for a type signature declaration:
+{- | Pattern for the function type signature declaration:
 
 @
 foo :: Some -> Type
 @
 -}
 typeSig :: PatternAst
 typeSig = PatternAstNode $ one ("TypeSig", "Sig")
+
+{- | @data@ or @newtype@ declaration.
+-}
+dataDecl :: PatternAst
+dataDecl = PatternAstNode $ one ("DataDecl", "TyClDecl")
+
+{- | Constructor of a plain data type or newtype. Children of node
+that matches this pattern are constructor fields.
+-}
+constructor :: PatternAst
+constructor = PatternAstNode $ one ("ConDeclH98", "ConDecl")
+
+{- | Lazy data type field. Comes in two shapes:
+
+1. Record field, like: @foo :: Text@
+2. Simple type: @Int@
+-}
+lazyField :: PatternAst
+lazyField = lazyRecordField ||| type_
+
+{- | Pattern for any occurrence of a plain type. Covers the following
+cases:
+
+* Simple type: Int, Bool, a
+* Higher-kinded type: Maybe Int, Either String a
+* Type in parenthesis: (Int)
```

Comment:
I s this really a separate case? :thinking:

Reply:
Yes, it is, and I was surprised by this as well :unamused: